exporter/dynatraceexporter/metrics_exporter.go

```diff
@@ -94,59 +110,67 @@ func (e *exporter) PushMetricsData(ctx context.Context, md pdata.Metrics) (dropp
-		}
-	}
-
-	return dropped, nil
+	return output, dropped
 }
 
-func (e *exporter) send(ctx context.Context, message string) error {
+func (e *exporter) send(ctx context.Context, message string) (int, error) {
 	e.logger.Debug("Sending lines to Dynatrace\n" + message)
 	req, err := http.NewRequestWithContext(ctx, "POST", e.cfg.Endpoint, bytes.NewBufferString(message))
 	if err != nil {
-		return fmt.Errorf("dynatrace error while creating HTTP request: %s", err.Error())
+		return 0, consumererror.Permanent(err)
 	}
 
-	req.Header.Add("Content-Type", "text/plain; charset=UTF-8")
-	req.Header.Add("Authorization", "Api-Token "+e.cfg.APIToken)
-	req.Header.Add("User-Agent", "opentelemetry-collector")
-
 	resp, err := e.client.Do(req)
 	if err != nil {
-		return fmt.Errorf("error sending HTTP request: %s", err.Error())
+		return 0, fmt.Errorf("error sending HTTP request: %w", err)
 	}
 	defer resp.Body.Close()
 
-	bodyBytes, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return fmt.Errorf("error while receiving HTTP response: %s", err.Error())
+	if resp.StatusCode == 202 {
+		// All metrics were accepted
+		return 0, nil
 	}
 
-	responseBody := metricsResponse{}
-	if err := json.Unmarshal(bodyBytes, &responseBody); err != nil {
-		e.logger.Error(fmt.Sprintf("failed to unmarshal response: %s", err.Error()))
-	} else {
-		e.logger.Debug(fmt.Sprintf("Exported %d lines to Dynatrace", responseBody.Ok))
-
-		if responseBody.Invalid > 0 {
-			e.logger.Debug(fmt.Sprintf("Failed to export %d lines to Dynatrace", responseBody.Invalid))
+	if resp.StatusCode == 400 || (resp.StatusCode >= 200 && resp.StatusCode <= 299) {
+		// At least some metrics were not accepted
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return 0, fmt.Errorf("error while receiving HTTP response: %s", err.Error())
 		}
 
-		if responseBody.Error != "" {
-			e.logger.Error(fmt.Sprintf("Error from Dynatrace: %s", responseBody.Error))
+		responseBody := metricsResponse{}
+		if err := json.Unmarshal(bodyBytes, &responseBody); err != nil {
+			e.logger.Error(fmt.Sprintf("failed to unmarshal response: %s", err.Error()))
+		} else {
+			e.logger.Debug(fmt.Sprintf("Exported %d lines to Dynatrace", responseBody.Ok))
+
+			if responseBody.Invalid > 0 {
+				e.logger.Error(fmt.Sprintf("%d lines rejected by Dynatrace", responseBody.Invalid))
+			}
+
+			if responseBody.Error != "" {
+				e.logger.Error(fmt.Sprintf("Error from Dynatrace: %s", responseBody.Error))
+			}
+
+			return responseBody.Invalid, nil
 		}
 	}
 
-	if !(resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusAccepted) {
-		return fmt.Errorf("request failed with response code: %d", resp.StatusCode)
+	if resp.StatusCode == 401 || resp.StatusCode == 403 || resp.StatusCode == 413 {
+		// Unauthorized and Unauthenticated errors are permanent
+		// If a payload is too large, resending it will not help
+		return 0, consumererror.Permanent(fmt.Errorf(resp.Status))
```

Comment:
Is this permanent per lifetime of the collector/exporter or per metrics batch?

Reply:
Just the batch. We could shutdown on a lifetime error like 401/403 if you prefer?

Reply:
Since they are unrecoverable I'd shut down the exporter.